fix(@mastra/core): resolve OpenAI reasoning model memory issue on sec…

roaminro opened the pull request:
## Description

…ond generate call

This commit addresses a bug where the second `agent.generate()` call with OpenAI reasoning models would fail due to missing `providerMetadata` for text parts. The fix ensures that `text-start` events correctly capture the `msg_` itemId, allowing for proper memory replay and preventing errors related to missing required items.

Additionally, comprehensive tests have been added to validate the fix and ensure that both reasoning and text parts retain their respective itemIds during interactions.

## Related Issue(s)

Fixes #11481.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Code refactoring
- [ ] Performance improvement
- [ ] Test update

## Checklist

- [ ] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Resolved memory replay mismatch by correctly capturing text-part provider metadata on stream start and clearing it on stream end, preventing ID leakage between reasoning and text parts.
  * Ensures generate and follow-up flows include distinct metadata for reasoning and text segments.

* **Tests**
  * Added unit and integration tests covering metadata capture/propagation across streaming, generate, tool-call, replay flows, and compatibility with the newer response-item reference format.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

roaminro committed:
fix(@mastra/core): resolve OpenAI reasoning model memory issue on second generate call

This commit addresses a bug where the second `agent.generate()` call with OpenAI reasoning models would fail due to missing `providerMetadata` for text parts. The fix ensures that `text-start` events correctly capture the `msg_` itemId, allowing for proper memory replay and preventing errors related to missing required items.

Additionally, comprehensive tests have been added to validate the fix and ensure that both reasoning and text parts retain their respective itemIds during interactions.

Fixes #11481.

abhiaiyer91 commented on packages/core/src/loop/workflows/agentic-execution/llm-execution-step.ts:
```diff
@@ -139,6 +139,23 @@ async function processOutputStream<OUTPUT extends OutputSchema = undefined>({
+        // preserved so that when memory is replayed, OpenAI sees the required
+        // following item for the reasoning part.
+        if (chunk.payload.providerMetadata) {
+          runState.setState({
```
text end will need to unset this then!

roaminro committed:
test: add test for text-end providerMetadata cleanup

Add test to verify that text-end properly clears providerOptions
to prevent text's providerMetadata from leaking into subsequent
parts like tool calls.

roaminro committed:
test: add comprehensive test for reasoning → text → tool call cleanup

Add test to verify the full cleanup cycle:
- reasoning-end clears reasoning providerMetadata (rs_xxx)
- text-start captures text providerMetadata (msg_xxx)
- text-end clears text providerMetadata

This ensures neither reasoning nor text metadata leaks into
subsequent parts like tool calls.

abhiaiyer91 committed:
Merge branch 'main' into roamin/fix-reasoning

roaminro committed:
Enhance agent memory tests to support v6 OpenAI Response API format

- Updated test logic to handle both legacy and v6 formats for agent memory requests.
- Added checks for item_reference messages in v6 format, allowing for proper validation of user and assistant message counts.
- Ensured that tool messages are correctly filtered out based on the response format.
- Improved overall test coverage for message structure and response validation.

roaminro committed:
coderabbit suggestions

roaminro committed:
Merge branch 'main' into roamin/fix-reasoning

roaminro committed:
Merge branch 'main' into roamin/fix-reasoning

roaminro committed:
Merge branch 'main' into roamin/fix-reasoning

roaminro committed:
Merge branch 'main' into roamin/fix-reasoning

roaminro committed:
Merge branch 'main' into roamin/fix-reasoning

roaminro committed:
Merge origin/main into roamin/fix-reasoning

Resolve conflict in agent-memory.ts by keeping v6 Response API format
handling that properly filters item_reference messages.
